No version set for command ... - still happening in v0.7.1

Still having the "no version set" issue (originally reported in https://github.com/asdf-vm/asdf/issues/488) with v0.7.1 even after run `asdf reshim`

### Steps to reproduce

my .tool-versions is

```
erlang 21.0.6
elixir path:/home/eksperimental/elixir/
```

```
$ cd /home/eksperimental/elixir/
$ iex
asdf: No version set for command iex
you might want to add one of the following in your .tool-versions file:

elixir 1.4.4
elixir 1.7.4
elixir 1.8.0
elixir 1.8.1
```

### Expected behavior

to run `/home/eksperimental/elixir/bin/iex`

### Actual behavior

```
asdf: No version set for command iex
```

### Environment

**OS**: Ubuntu GNU/Linux 16.04 32 bits

**asdf version**: v0.7.1

I updated from v0.5.1 to v0.7.1 today on macOS. Now I'm seeing this issue:

```
asdf: No version set for command erl
you might want to add one of the following in your .tool-versions file:

erlang 19.3
erlang 20.0
erlang 20.1
erlang 21.1
erlang 21.2.3
```

I had a similar issue today with a Linux computer. I fixed it by uninstalling all the versions and reinstalling them. I did that on macOS. Unfortunately, it's no longer able to install Erlang. 

```
Building Erlang/OTP 21.2.3 (asdf_21.2.3), please wait...
APPLICATIONS DISABLED (See: /Users/robert/.asdf/plugins/erlang/kerl-home/builds/asdf_21.2.3/otp_build_21.2.3.log)
 * crypto         : No usable OpenSSL found
 * odbc           : ODBC library - header check failed
 * ssh            : No usable OpenSSL found
 * ssl            : No usable OpenSSL found

DOCUMENTATION INFORMATION (See: /Users/robert/.asdf/plugins/erlang/kerl-home/builds/asdf_21.2.3/otp_build_21.2.3.log)
 * documentation  : 
 *                  fop is missing.
 *                  Using fakefop to generate placeholder PDF files.

Build failed.
      _driver_dl_open in io.o
      _driver_dl_sym in io.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[4]: *** [/Users/robert/.asdf/plugins/erlang/kerl-home/builds/asdf_21.2.3/otp_src_21.2.3/bin/x86_64-apple-darwin18.5.0/beam.prof.smp] Error 1
make[3]: *** [opt] Error 2
make[2]: *** [opt] Error 2
make[1]: *** [smp] Error 2
make: *** [emulator_profile_generate] Error 2

Please see /Users/robert/.asdf/plugins/erlang/kerl-home/builds/asdf_21.2.3/otp_build_21.2.3.log for full details.
```

Here's the log:

[otp_build_21.2.3.log](https://github.com/asdf-vm/asdf/files/3188979/otp_build_21.2.3.log)

--- a/lib/commands/reshim.sh
+++ b/lib/commands/reshim.sh
@@ -65,7 +65,7 @@ write_shim_script() {
   shim_path="$(asdf_data_dir)/shims/$executable_name"
 
   if [ -f "$shim_path" ]; then
-    if ! grep "# asdf-plugin: ${plugin_name} ${version}" "$shim_path" >/dev/null; then
+    if ! grep -x "# asdf-plugin: ${plugin_name} ${version}" "$shim_path" >/dev/null; then
      sed -i.bak -e "s/exec /# asdf-plugin: ${plugin_name} ${version}\\"$'\n''exec /' "$shim_path"
      rm "$shim_path".bak
     fi

--- a/lib/commands/reshim.sh
+++ b/lib/commands/reshim.sh
@@ -120,7 +120,7 @@ remove_obsolete_shims() {
   obsolete_shims=$(comm -23 <(echo "$shims") <(echo "$exec_names"))
 
   for shim_name in $obsolete_shims; do
-    remove_shim_for_version "$plugin_name" "$version" "$shim_name"
+    remove_shim_for_version "$plugin_name" "$full_version" "$shim_name"
   done
 }
 
@@ -137,11 +137,11 @@ remove_shim_for_version() {
   local count_installed
   count_installed=$(list_installed_versions "$plugin_name" | wc -l)
 
-  if ! grep "# asdf-plugin: $plugin_name $version" "$shim_path" > /dev/null 2>&1; then
+  if ! grep -x "# asdf-plugin: $plugin_name $version" "$shim_path" > /dev/null 2>&1; then
     return 0
   fi
 
-  sed -i.bak -e "/# asdf-plugin: $plugin_name $version/d" "$shim_path"
+  sed -i.bak -e "/# asdf-plugin: $plugin_name $version"'$/d' "$shim_path"
   rm "$shim_path".bak
 
   if ! grep "# asdf-plugin:" "$shim_path" > /dev/null || \
@@ -154,6 +154,6 @@ remove_shims_for_version() {
   local plugin_name=$1
   local full_version=$2
   for shim_path in $(plugin_shims "$plugin_name" "$full_version"); do
-    remove_shim_for_version "$plugin_name" "$version" "$shim_path"
+    remove_shim_for_version "$plugin_name" "$full_version" "$shim_path"
   done
 }

--- a/lib/utils.sh
+++ b/lib/utils.sh
@@ -545,7 +545,7 @@ is_executable() {
 plugin_shims() {
   local plugin_name=$1
   local full_version=$2
-  grep -l "# asdf-plugin: $plugin_name $full_version" "$(asdf_data_dir)/shims"/* 2>/dev/null
+  grep -lx "# asdf-plugin: $plugin_name $full_version" "$(asdf_data_dir)/shims"/* 2>/dev/null
 }
 
 shim_plugin_versions() {

--- a/test/reshim_command.bats
+++ b/test/reshim_command.bats
@@ -57,6 +57,17 @@ teardown() {
   [ "$status" -eq 0 ]
 }
 
+@test "reshim should not remove metadata of removed prefix versions" {
+  run asdf install dummy 1.0
+  run asdf install dummy 1.0.1
+  run rm "$ASDF_DIR/installs/dummy/1.0/bin/dummy"
+  run asdf reshim dummy
+  [ "$status" -eq 0 ]
+  [ -f "$ASDF_DIR/shims/dummy" ]
+  run grep "asdf-plugin: dummy 1.0.1" "$ASDF_DIR/shims/dummy"
+  [ "$status" -eq 0 ]
+}
+
 @test "reshim should not duplicate shims" {
   cd $PROJECT_DIR
 

--- a/test/reshim_command.bats
+++ b/test/reshim_command.bats
@@ -15,6 +15,19 @@ teardown() {
 }
 
 
+@test "reshim should allow prefixes of other versions" {
+  run asdf install dummy 1.0.1
+  run asdf install dummy 1.0
+
+  run asdf reshim
+  [ "$status" -eq 0 ]
+
+  run grep "asdf-plugin: dummy 1.0.1" "$ASDF_DIR/shims/dummy"
+  [ "$status" -eq 0 ]
+  run grep "asdf-plugin: dummy 1.0\$" "$ASDF_DIR/shims/dummy"
+  [ "$status" -eq 0 ]
+}
+
 @test "reshim command should remove shims of removed binaries" {
   run asdf install dummy 1.0
   [ "$status" -eq 0 ]

